sos/deviceauth.py

```diff
@@ -0,0 +1,291 @@
+        return None
+    except Exception as e:
+        logger.error("Error encoutered while accessing keyring. {}".format(e))
+        return None
+    if key_username.data.decode() != os.getlogin():
+        return None
+
+    return key_refresh_token.data.decode()
+
+
+class AuthClass:
+    """
+    Device Authorization Class
+    """
+
+    def __init__(self):
+
+        self.client_identifier_url = "https://sso.redhat.com/auth/"\
+            "realms/redhat-external/protocol/openid-connect/auth/device"
+        self.token_endpoint = "https://sso.redhat.com/auth/realms/"\
+            "redhat-external/protocol/openid-connect/token"
+        self._access_token = None
+        self._access_expires_at = None
+        self._refresh_token = None
+        self._refresh_expires_at = None
+        self._refresh_expires_in = None
+        self._user_verification_url = None
+        self.__device_code = None
+        self.RHELKey = Key()
+
+        # Lets check first if we have keyctl installed so we can
+        # store the token in the keyring
+        if not shutil.which("keyctl"):
+            raise Exception("keyctl tool is not installed"
+                            " and is required to store auth tokens.")
+
+        self._use_device_code_grant()
+
+    def _use_device_code_grant(self):
+        """
+        Start the device auth flow. First check for the refresh token stored in
+        the session keyring. If they are not stored
+        or are invalid, request new device code. If the stored refresh token is
+        valid, use it to get new access_token
+
+        """
+        stored_refresh_token = try_read_refresh_token()
+
+        if not stored_refresh_token:
+            self._request_device_code()
+            print(
+                "Please visit the following URL to authenticate this"
+                f" device {self._verification_uri_complete}")
+            self.poll_for_auth_completion()
+        else:
+            self._use_refresh_token_grant(stored_refresh_token)
+
+    def _request_device_code(self):
+        """
+        Initialize new Device Authorization Grant attempt by
+        requesting a new device code.
+
+        """
+        data = "client_id={}".format(DEVICE_AUTH_CLIENT_ID)
+        headers = {'content-type': 'application/x-www-form-urlencoded'}
+        try:
+            res = requests.post(
+                self.client_identifier_url,
+                data=data,
+                headers=headers)
+            res.raise_for_status()
+            response = res.json()
+            self._user_code = response.get("user_code")
+            self._verification_uri = response.get("verification_uri")
+            self._interval = response.get("interval")
+            self.__device_code = response.get("device_code")
+            self._verification_uri_complete = response.get(
+                "verification_uri_complete")
+        except requests.HTTPError as e:
+            raise e
```

Comment:
Maybe add response status code to be more informative?

Reply:
ack, good idea!